connector.go

```diff
@@ -0,0 +1,218 @@
+	channel      *amqp.Channel
+	opt          *ConnectorOptions
+	closeWG      *sync.WaitGroup
+	logger       *log
+	instanceType string
+}
+
+func connect(params *connectParams) (*amqp.Connection, *amqp.Channel, error) {
+	const errMessage = "failed to connect to rabbitmq: %w"
+
+	conn := params.conn
+	channel := params.channel
+	var err error
+
+	if params.conn == nil {
+		conn, err = amqp.DialConfig(params.opt.uri, amqp.Config(*params.opt.Config))
+		if err != nil {
+			return nil, nil, fmt.Errorf(errMessage, err)
+		}
+
+		go watchConnectionNotifications(conn, publish, params.closeWG, params.logger)
+
+		channel, err = createChannel(conn, params.opt.PrefetchCount)
+		if err != nil {
+			return nil, nil, fmt.Errorf(errMessage, err)
+		}
+
+		go watchChannelNotifications(channel, publish, params.opt.ReturnHandler, params.closeWG, params.logger)
+	}
+
+	return conn, channel, nil
+}
+
+func createChannel(conn *amqp.Connection, prefetchCount int) (*amqp.Channel, error) {
+	const errMessage = "failed to create channel: %w"
+
+	channel, err := conn.Channel()
+	if err != nil {
+		return nil, fmt.Errorf(errMessage, err)
+	}
+
+	if prefetchCount > 0 {
+		err = channel.Qos(prefetchCount, 0, false)
+		if err != nil {
+			return nil, fmt.Errorf(errMessage, err)
+		}
+	}
+
+	return channel, nil
+}
+
+func watchConnectionNotifications(conn *amqp.Connection, name string, closeWG *sync.WaitGroup, logger *log) {
+	for {
+		select {
+		case err := <-conn.NotifyClose(make(chan *amqp.Error)):
+			if err == nil {
+				slog.Debug("closed connection", "type", name)
+
+				closeWG.Done()
+
+				return
+			}
+
+			//nolint: godox // follow-up task
+			// TODO: reconnect logic
+			logger.logDebug("connection unexpectedly closed", "type", name, "cause", err)
+
+		case block := <-conn.NotifyBlocked(make(chan amqp.Blocking)):
+			logger.logWarn("connection exception", "type", name, "cause", block.Reason)
+		}
+	}
+}
+
+func watchChannelNotifications(channel *amqp.Channel, name string, returnHandler ReturnHandler, closeWG *sync.WaitGroup, logger *log) {
+	for {
+		select {
+		case err := <-channel.NotifyClose(make(chan *amqp.Error)):
+			if err == nil {
+				slog.Debug("closed channel", "type", name)
+
+				closeWG.Done()
+
+				return
+			}
+
+			//nolint: godox // follow-up task
+			// TODO: reconnect logic
+			logger.logDebug("channel unexpectedly closed", "type", name, "cause", err)
+
+		case tag := <-channel.NotifyCancel(make(chan string)):
+			logger.logWarn("cancel exception", "type", name, "cause", tag)
+
+		case rtrn := <-channel.NotifyReturn(make(chan amqp.Return)):
+			if returnHandler != nil {
+				returnHandler(Return(rtrn))
+
+				continue
+			}
+
+			logger.logWarn(
+				"message could not be published",
+				"replyCode", rtrn.ReplyCode,
+				"replyText", rtrn.ReplyText,
+				"messageId", rtrn.MessageId,
+				"correlationId", rtrn.CorrelationId,
+				"exchange", rtrn.Exchange,
+				"routingKey", rtrn.RoutingKey,
+			)
+		}
+	}
+}
```

Comment:
why are the channels now inlined in here? 😄 why not just start the for loop in a go routine and keep everything else as it was